### PR TITLE
Add more CI

### PR DESCRIPTION
**Pull request by neersighted**
* Add commitlint (with Conventional Commits style)
* Add building of Nebula
* Better artifact naming
* Related fixes: POSIX helpers, database-clean invocation

### CODE REVIEW

File: .github/workflows/ci-cd.yml [hunk 1]
```diff
@@ -7,61 +7,155 @@ on:
   pull_request:
 
 jobs:
+  setup:
+    runs-on: ubuntu-latest
+    outputs:
+      short_sha: ${{ steps.sha.outputs.short_sha }}
+      go_version: ${{ steps.read.outputs.go_version }}
+      msys2_packages: ${{ steps.read.outputs.msys2_packages }}
+    steps:
```
Comment (ngld) [inline .github/workflows/ci-cd.yml]: Using a separate step to read this stuff once before the matrix steps is a really neat solution.

File: .github/workflows/ci-cd.yml [hunk 2]
```diff
@@ -7,61 +7,155 @@ on:
...
+          go-version: ${{ needs.setup.outputs.go_version }}
+
+      - name: Cache Go dependencies
+        uses: actions/cache@v2
+        with:
+          key: golang-${{ runner.os }}-${{ hashFiles('**/go.sum') }}
+          restore-keys: golang-${{ runner.os }}-
+          path: |
+            ~/go/pkg
+            ~/.cache/go-build
```
Comment (ngld) [inline .github/workflows/ci-cd.yml:55]: I'm not sure how effective this will be since the cache is only updated if the key didn't match at the start. This means that the build cache won't be updated until the Go dependencies change. However, if caching the compiled dependencies reduces build time enough, this could be useful.

File: .github/workflows/ci-cd.yml [hunk 3]
```diff
@@ -7,61 +7,155 @@ on:
...
+
+      - name: Setup Go toolchain
+        uses: actions/setup-go@v2
+        with:
+          go-version: ${{ needs.setup.outputs.go_version }}
+
+      - name: Cache Go dependencies
+        uses: actions/cache@v2
+        with:
+          key: golang-${{ runner.os }}-${{ hashFiles('**/go.sum') }}
```
Comment (ngld) [inline .github/workflows/ci-cd.yml]: Won't this cause Nebula and Knossos to share a Go cache on Linux although they use different dependencies?

File: .github/workflows/ci-cd.yml [hunk 4]
```diff
@@ -7,61 +7,155 @@ on:
...
+        uses: actions/upload-artifact@v2
+        with:
+          name: Nebula-server-${{ needs.setup.outputs.short_sha }}
+          path: build/nebula
+
+      - name: Upload frontend artifacts
+        uses: actions/upload-artifact@v2
+        with:
+          name: Nebula-frontend-${{ needs.setup.outputs.short_sha }}
+          path: packages/front/dist/prod
```
Comment (ngld) [inline .github/workflows/ci-cd.yml]: I think these steps are redundant since users will never run Nebula on their own PCs and I'd rather use a different distribution method for the production server.
It makes sense for Knossos because it allows users to download and use the latest (unreleased) version.

File: .github/workflows/ci-cd.yml [hunk 5]
```diff
@@ -70,17 +164,19 @@ jobs:
...
         with:
           key: third_party-${{ runner.os }}-${{ hashFiles('packages/build-tools/DEPS.yml') }}
           restore-keys: third_party-${{ runner.os }}-
           path: |
             packages/build-tools/DEPS.stamps
             third_party
 
+      - name: Configure build
+        run: ${{ matrix.script }} configure build=Release ${{ env.extra_options }}
+
```
Comment (ngld) [inline .github/workflows/ci-cd.yml]: Why did you move the configure step? Nothing in the "Fetch system dependencies" step depends on the configure result and more importantly if we decide to search for static libraries on Linux, the configure step won't find them if we run it before the libraries are installed.

File: packages/build-tools/cmd/main.go [hunk 6]
```diff
@@ -18,5 +18,5 @@ func init() {
 }
 
 func Execute() {
-	cobra.CheckErr(rootCmd.Execute())
+	rootCmd.Execute()
```
Comment (ngld) [inline packages/build-tools/cmd/main.go]: Why did you remove the error check? This essentially silences errors on *all* tool commands in *all* cases. This *does* seemingly fix `tool rm -f ...` but that hopefully isn't what you were trying to do with this change. It also doesn't actually fix `tool rm -f` since an error will still cause it to abort removing files which means parameters after a missing file won't be deleted.

File: packages/build-tools/cmd/posix_helpers.go [hunk 7]
```diff
@@ -114,18 +114,18 @@ var rmCmd = &cobra.Command{
...
+				}
+			} else if info.IsDir() && !recursive {
 				return eris.Errorf("%s is a directory but -r wasn't passed", item)
 			}
 		}
 
 		for _, item := range items {
 			err := os.RemoveAll(item)
-			if err != nil && (!force || !eris.Is(err, os.ErrNotExist)) {
+			if err != nil {
```
Comment (ngld) [inline packages/build-tools/cmd/posix_helpers.go:128]: You want to keep this check since missing files aren't removed from `items` during the first iteration and thus are still present during this iteration.

File: packages/build-tools/cmd/posix_helpers.go [hunk 8]
```diff
@@ -114,18 +114,18 @@ var rmCmd = &cobra.Command{
 
 		for _, item := range items {
 			info, err := os.Stat(item)
-			if err != nil && !force {
-				return eris.Wrapf(err, "Could not stat %s", item)
-			}
-
-			if info.IsDir() && !recursive {
+			if err != nil {
+				if !(force && eris.Is(err, os.ErrNotExist)) {
```
Comment (ngld) [inline packages/build-tools/cmd/posix_helpers.go]: I think `!force || !eris.Is(err, os.ErrNotExist)` would be easier to read. Can you change this?

**Review by ngld — CHANGES_REQUESTED** [review]
Fun fact: I used to have commitlint as a pre-commit hook but eventually removed it because it complicated setup and it was slow enough to be annoying.
I still think a pre-commit hook would be a better solution because it would provide much more immediate feedback (fixing commit messages with an interactive rebase is annoying) but I couldn't find a better solution (launching anything Node.js based through Yarn 2 will take longer than I'd like for something simple like this).

I'm not sure why we would want to provide downloads for Nebula. Who would use them?

The issue with Go caching can be resolved by either using two separate cache keys or installing all Go dependencies regardless of which package is built. The former might be much easier than the latter solution.

Finally, I'd like to skip the Nebula job if Nebula's source didn't change and the Knossos jobs in case Knossos' source didn't change but I'm not sure if GitHub supports that. We could implement a check in a shell script which might work if we run that check in `setup` and then use that result in a condition for the jobs.

File: .github/workflows/ci-cd.yml (hunk 1 above)
Comment (neersighted) [reply .github/workflows/ci-cd.yml]: Yeah. it seemed pretty obvious in hindsight. I did not know `needs` was a thing before -- GH Actions is really best in class right now.

File: .github/workflows/ci-cd.yml (hunk 2 above)
Comment (neersighted) [reply .github/workflows/ci-cd.yml:55]: I got this from `golangci-lint` and it seemed like there was no downside. It will only cache on the first compile with a set of deps, but that seems entirely acceptable.

File: .github/workflows/ci-cd.yml (hunk 5 above)
Comment (neersighted) [reply .github/workflows/ci-cd.yml]: I think that may have been an accident. Will swap it back.

File: packages/build-tools/cmd/main.go (hunk 6 above)
Comment (neersighted) [reply packages/build-tools/cmd/main.go]: So unless I'm very mistaken, there's already an implicit `CheckErr` in `Execute`. Before this change, an error would result in the following:

```
Error: error message
Usage: help message for the tool
...
...
Error: error message
```

After this change, the duplicate second error message is dropped, so it's just `Error:`, `Usage:`. Am I missing a case where this is necessary to print the error message?

File: packages/build-tools/cmd/posix_helpers.go (hunk 7 above)
Comment (neersighted) [reply packages/build-tools/cmd/posix_helpers.go:128]: `RemoveAll` actually does not throw `os.ErrNotExist` as it is essentially `rm -f` in syscall form. So there is no need to handle this down below.

File: packages/build-tools/cmd/posix_helpers.go (hunk 8 above)
Comment (neersighted) [reply packages/build-tools/cmd/posix_helpers.go]: Yeah that's probably a bit more understandable.

**Comment by neersighted** [comment]
It should be easy enough to only run on certain files changed. What I'm more concerned with is how to keep that list -- ideally I think the build system could be asked? Let me know if you have thoughts on that detail.

**Comment by neersighted** [comment]
I think my idea with the Nebula downloads was that it would be easy to test PRs locally, but in hindsight clone the PR and build it yourself is faster and easier.

File: .github/workflows/ci-cd.yml (hunk 2 above)
Comment (ngld) [reply .github/workflows/ci-cd.yml:55]: The downside is that cache space is limited to 5 GiB. However, I think we're only using about 1 GiB right now. With a few open PRs which change dependencies that could grow to 2-3 GiB but that still seems safe.
Unfortunately, there doesn't seem to be a way to check currently used cache space. I just look at the reported cache size in the cache restore steps.

File: packages/build-tools/cmd/main.go (hunk 6 above)
Comment (ngld) [reply packages/build-tools/cmd/main.go]: Without `CheckErr()`, the process always exits with code 0. If you just want to remove the duplicate error message, you should change that call to the following:
```go
err := rootCmd.Execute()
if err != nil {
  os.Exit(1)
}
```

**Comment by ngld** [comment]
I've fixed the exit code handling. I have an idea regarding the changed files but I'll do that in a separate PR.

**Review by ngld — APPROVED** [review]
